File: src/brain_teasers/python/k_closest_points/src/k_closest_points.py
"""Find the K closest points to the origin."""

from __future__ import annotations

import math


def k_closest_points(
    points: list[tuple[int, int]], k: int
) -> list[tuple[int, int]]:
    """
    Return the K closest points to the origin (0, 0).

    Uses Euclidean distance to measure closeness.

    Args:
        points: List of (x, y) coordinate tuples.
        k: Number of closest points to return.

    Returns:
        List of the K closest points to the origin.
    """
    # Create a list of (distance, point) tuples
    distances = [(math.sqrt(x**2 + y**2), (x, y)) for x, y in points]
    # Sort by distance
    distances.sort()
    # Return the K closest points
    return [point for _, point in distances[:k]]

File: src/brain_teasers/cpp/k_closest_points/src/k_closest_points.cpp
#include "k_closest_points.h"
#include <algorithm>
#include <queue>

// Helper function to calculate squared distance from origin
static int squaredDistance(const Point& p) {
    return p.x * p.x + p.y * p.y;
}

std::vector<Point> kClosestPoints(std::vector<Point>& points, int k) {
    if (k <= 0) return {};
    if (k >= static_cast<int>(points.size())) return points;
    
    // Use partial_sort for O(n log k) complexity
    std::partial_sort(points.begin(), points.begin() + k, points.end(),
        [](const Point& a, const Point& b) {
            return squaredDistance(a) < squaredDistance(b);
        });
    
    return std::vector<Point>(points.begin(), points.begin() + k);
}
